feat(core): add support for @HostProperty and @HostListener

vsavkin opened the pull request:
Example:

```
@Directive({selector: 'my-directive'})
class MyDirective {
  @HostProperty("attr.my-attr") myAttr: string;
  @HostListener("click", ["$event.target"])
  onClick(target) {
    this.target = target;
  }
}
```

mhevery commented on modules/angular2/src/core/metadata/directives.ts:
```diff
@@ -1109,4 +1109,43 @@ export class PropertyMetadata {
+/**
+ * Declare a host property.
+ *
+ * ## Example
```
put some text that this is equivalent to the `Directive.host` syntax.

mhevery commented on modules/angular2/src/core/metadata/directives.ts:
```diff
@@ -1109,4 +1109,43 @@ export class PropertyMetadata {
+
+/**
+ * Declare a host listener.
+ *
```
same here

mhevery commented on modules/angular2/test/core/compiler/integration_spec.ts:
```diff
@@ -2203,8 +2243,16 @@ class DirectiveThrowingAnError {
   @Property("elProp") dirProp: string;
   @Event('elEvent') event = new EventEmitter();
 
+  @HostProperty("attr.my-attr") myAttr: string;
```
rename to `HostBindings`

vsavkin committed:
feat(core): add support for @HostBinding and @HostListener

Example:

@Directive({selector: 'my-directive'})
class MyDirective {
  @HostBinding("attr.my-attr") myAttr: string;
  @HostListener("click", ["$event.target"])
  onClick(target) {
    this.target = target;
  }
}

mary-poppins commented:
Merging PR #3996 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-3996](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-3996).
